devicetree: Add DT_FIXED_PARTITION_ADDR macro

57300 opened the pull request:
This convenience API returns the absolute address of a fixed partition, i.e., relative offset + base address. It's distinct from `DT_REG_ADDR()` and `FIXED_PARTITION_OFFSET()`, both of which return just the offset.

The base address is taken from the parent memory node as given by the newly added `DT_MEM_FROM_FIXED_PARTITION()`. This is expected to ensure that the returned address is directly addressable by the CPU. This is also meant to prevent `DT_FIXED_PARTITION_ADDR()` from working with external memory partitions.

57300 committed:
devicetree: Add DT_FIXED_PARTITION_ADDR macro

This convenience API returns the absolute address of a fixed partition,
i.e., relative offset + base address. It's distinct from `DT_REG_ADDR()`
and `FIXED_PARTITION_OFFSET()`, both of which return just the offset.

The base address is taken from the parent memory node as given by the
newly added `DT_MEM_FROM_FIXED_PARTITION()`. This is expected to ensure
that the returned address is directly addressable by the CPU. This is
also meant to prevent `DT_FIXED_PARTITION_ADDR()` from working with
external memory partitions.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
doc: release: 3.5: Add note on new fixed partition API

List `DT_MEM_FROM_FIXED_PARTITION()` and `DT_FIXED_PARTITION_ADDR()` as
the newly added macros.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

Laczen commented:
@57300, what would this macro be used for ?

57300 commented:
> @57300, what would this macro be used for ?

It would be used to reference a partition in memory-mapped flash by its physical address, so that we could e.g., read its contents directly (where a flash driver is not required) or execute code from it. This could be used in boot code, linker scripts, anything really.

So far, I've identified uses for it downstream.

Laczen commented:
@57300, doesn't the macro return the same address for internal flash residing at 0x0 and external flash ?

57300 commented:
> @57300, doesn't the macro return the same address for internal flash residing at 0x0 and external flash ?

Correct, and I realize now that this can be a problem. I will rewrite the macro to disallow its use with external flash.

57300 committed:
devicetree: Add DT_MEM_FROM_FIXED_PARTITION macro

This is a new fixed-partitions API, which serves as a companion to the
existing `DT_MTD_FROM_FIXED_PARTITION()`. In the following example:

  flash-controller@0 {
          compatible = "flash-controller";
          flash@1000000 {
                  compatible = "soc-nv-flash";
                  partitions {
                          compatible = "fixed-partitions";
                          partition@3a000 {};
                  };
          };
  };

`DT_MTD_FROM_FIXED_PARTITION()` would let us map `partition@3a000` to
`flash-controller@0`. Now, the new `DT_MEM_FROM_FIXED_PARTITION()` can
let us retrieve the memory node (`flash@1000000`) as well, in a manner
consistent with the existing API.

Caution: if a fixed partition is not said to belong to a memory node,
like in this alternative example:

  flash@0 {
          compatible = "spi-nor";
          partitions {
                  compatible = "fixed-partitions";
                  partition@0 {};
          };
  };

then `DT_MEM_FROM_FIXED_PARTITION()` will map `partition@0` to an
invalid node identifier. This partition belongs to `flash@0`, which
could be an MTD on a SPI bus.

For consistency, the existing `DT_MTD_FROM_FIXED_PARTITION()` is now re-
expressed in terms of the new `DT_MEM_FROM_FIXED_PARTITION()`.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 commented:
I added `DT_MEM_FROM_FIXED_PARTITION()` and rewrote `DT_FIXED_PARTITION_ADDR()` in terms of it, to catch unintended use with external flash.

@Laczen FYI

mbolivar-ampere reviewed (changes requested):
All new devicetree macros require coverage in tests/lib/devicetree/api; please add this.

@de-nordic are you OK with this?

57300 committed:
tests: lib: devicetree: api: Add test_fixed_partitions

Add test suite for the DT fixed-partitions API. It is verified on two
kinds of MTD nodes, one of which is meant to represent external memory.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 commented:
> All new devicetree macros require coverage in tests/lib/devicetree/api; please add this.

Rebased and added tests.

mbolivar-ampere reviewed (approved):
No objections but I think @de-nordic should approve

de-nordic commented on include/zephyr/devicetree/fixed-partitions.h:
```diff
@@ -78,15 +78,60 @@ extern "C" {
+ * @return the node identifier of the internal memory that contains the
+ * fixed-partitions node, or @ref DT_INVALID_NODE if it doesn't exist.
+ */
+#define DT_MEM_FROM_FIXED_PARTITION(node_id)                                                       \
```
I do not like the name, just briefly looking at it I thought it is intended to get some memory mapping, region, or device property  and the description at line 81 got me even more baffled.
Can we have something like `DT_MEM_DEV_NODE_ID_FROM_FIXED_PARTITION` ? Or something.

carlescufi replied:
@de-nordic it is consistent with the existing `DT_MTD_FROM_FIXED_PARTITION`

de-nordic reviewed (approved):
Looks OK.
But... we should log an issue to Flash Map API too look at as the Flash Map has not been taking the flash node address into account when calculating offset, which may be a problem.
Currently the flash@0 and flash@something, with compatible = "soc-nv-flash", will have the same device used for access and offsets of all partitions will be calculated as starting at 0, which means that partitions  defined at flash@something may not be correctly accessed with use of Flash Map, as flash controller will not automatically add "something" to offsets of partitions defined in flash map.
